### DISCUSSION TITLE
My android demo has been achieved with Filament, but we have encountered a memory problem. Can you give some suggestions or ideas?

### DISCUSSION THREAD

**Discussion by caorm** [google/filament#5631] [General]
（1）The filament android gltf demo effect, I changed it to a 3d model ：

Instructions：The model uses 1024*1024, and the hub uses 512*512. Others use 256*256。

（2）I use filament to implement this demo, but I found that the memory growth is relatively large：
（a）Android: google pixel 2xl：Increase 259M
                         Huawei mobile phone P30：Increase 148M

（b） IOS:iPhone 6s：Increase 80M
                  iPhone 8P 13.4.1   Increase 150M

So I found that the memory growth is related to the resolution of different mobile phones。

（3）also, The filament demo I changed also supports multiple models and scrollview scrolling switching effects。 because the effect requires multiple trolleys to slide and switch horizontally, my implementation plan : a trolley corresponds to a view, that is, corresponds to an engine (Engine::create((Engine::Backend)backend)) , which will cause the memory to increase several times. I tried to share one engine object (Engine, MaterialProvider, AssetLoader). Although it reduces some memory, it is still not enough.

（4）I also found that textures and resources cannot be shared among multiple Views, and if multiple models are used unreasonably, they will be overwritten, and only multiple ResourceLoader objects can be created.
![car4](https://user-images.githubusercontent.com/5361593/171407311-089e82b5-ca22-4a4e-95d6-888fe18bb3ab.png)

（5）Can you guys give some memory optimization suggestions？thank you very much！

**Comment by romainguy** [comment]
Memory usage will increase with resolution, and with the number of `SurfaceView` or `TextureView` you use. You should absolutely share the `Engine` and everything else you can across multiple views (meshes, textures, etc.). This is the best way to save memory. Also try to use as few (or better a single) `SurfaceView` or `TextureView` as each is triple-buffered by Android which will greatly increase memory usage.

Here is an example of an app I wrote that shares as much as it can across multiple items in a scrolling list: https://github.com/romainguy/sample-materials-shop/\

@prideout can you comment on this please:
> 4）I also found that textures and resources cannot be shared among multiple Views, and if multiple models are used unreasonably, they will be overwritten, and only multiple ResourceLoader objects can be created.

**Comment by prideout** [comment]
Hi, thanks for the questions!  Please note that we have a discussions tab.  Some thoughts:

(1) It is possible to persist a single ResourceLoader across multiple loads, and in fact that's what we do in `ModelViewer.kt`.

(2) If you have several instances of a single glTF asset, you should definitely use the existing instancing feature in glTF, which allows textures and resources to be shared amongst instances.  You can see `gltf_instances.cpp` for an example.

(3) Your code screenshot shows off a known limitation, which is that it can perform texture decoding for only a single glTF asset at a time.  Usually, a single glTF asset has enough textures to keep all your cores busy, so it probably wouldn't help load time to add more jobs.

(4) I don't understand the statement that "textures and resources cannot be shared among multiple Views" since Views contain references to Scenes (not textures and resources) and a single Scene can definitely be shared amongst multiple Views.
